main/EMC2302.c

```diff
@@ -0,0 +1,89 @@
+#include "driver/i2c.h"
+#include "esp_log.h"
+#include <stdio.h>
+
+#include "EMC2302.h"
+
+#define I2C_MASTER_SCL_IO 48 /*!< GPIO number used for I2C master clock */
+#define I2C_MASTER_SDA_IO 47 /*!< GPIO number used for I2C master data  */
+#define I2C_MASTER_NUM                                                                                                             \
+    0 /*!< I2C master i2c port number, the number of i2c peripheral interfaces available will depend on the chip */
+#define I2C_MASTER_FREQ_HZ 400000   /*!< I2C master clock frequency */
+#define I2C_MASTER_TX_BUF_DISABLE 0 /*!< I2C master doesn't need buffer */
+#define I2C_MASTER_RX_BUF_DISABLE 0 /*!< I2C master doesn't need buffer */
+#define I2C_MASTER_TIMEOUT_MS 1000
+
+static const char *TAG = "EMC2302.c";
+
+/**
+ * @brief Read a sequence of I2C bytes
+ */
+static esp_err_t register_read(uint8_t reg_addr, uint8_t * data, size_t len)
+{
+    return i2c_master_write_read_device(I2C_MASTER_NUM, EMC2302_I2CADDR_DEFAULT, &reg_addr, 1, data, len,
+                                        I2C_MASTER_TIMEOUT_MS / portTICK_PERIOD_MS);
+}
+
+/**
+ * @brief Write a byte to a I2C register
+ */
+static esp_err_t register_write_byte(uint8_t reg_addr, uint8_t data)
+{
+    int ret;
+    uint8_t write_buf[2] = {reg_addr, data};
+
+    ret = i2c_master_write_to_device(I2C_MASTER_NUM, EMC2302_I2CADDR_DEFAULT, write_buf, sizeof(write_buf),
+                                     I2C_MASTER_TIMEOUT_MS / portTICK_PERIOD_MS);
+
+    return ret;
+}
+
+// run this first. sets up the PWM polarity register
+void EMC2302_init(bool invertPolarity)
+{
+    if (invertPolarity) {
+        ESP_ERROR_CHECK(register_write_byte(EMC2302_PWM_POLARITY, 0b00011111));
+    }
+}
+
+// Sets the fan speed to a given percent
+void EMC2302_set_fan_speed(uint8_t devicenum, float percent)
+{
+    uint8_t speed;
+	uint8_t FAN_SETTING_REG = EMC2302_FAN1_SETTING + (devicenum * 0x10);
```

Comment:
Don't do arithmetic with constants, this is bound to cause problems. Use ifs or switch and handle out of bound devicenum.